src/za/ac/sun/cs/green/service/simplify/ConstantPropogation.java

```diff
@@ -0,0 +1,308 @@
+
+   public ConstantPropogation(Green solver) {
+       super(solver);
+   }
+
+   @Override
+   public Set<Instance> processRequest(Instance instance) {
+       @SuppressWarnings("unchecked")
+       Set<Instance> result = (Set<Instance>) instance.getData(getClass());
+       if (result == null) {
+           final Map<Variable, Variable> map = new HashMap<Variable, Variable>();
+           final Expression e = propogateConstants(instance.getFullExpression(), map);
+           final Instance i = new Instance(getSolver(), instance.getSource(), null, e);
+           result = Collections.singleton(i);
+           instance.setData(getClass(), result);
+       }
+       return result;
+   }
+
+   @Override
+	public void report(Reporter reporter) {
+		reporter.report(getClass().getSimpleName(), "invocations = " + invocations);
+   }
+public Expression propogateConstants(Expression expression, Map<Variable, Variable> map) {
+               //Use the visitor to propogate constants
+               try {
+                    log.log(Level.FINEST, "Before ConstantPropogation: " + expression);
+       			    invocations++;
+
+                    HashMap <Variable, Constant> var_map;
+
+                    MapVisitor mapVisitor = new MapVisitor();
+        			expression.accept(mapVisitor);
+        			var_map = mapVisitor.getMap();
+                    //log.log(Level.FINEST, "\n__getMap__\n " +
+                    //mapVisitor.getLogStr());
+
+                    ConstantPropogationVisitor propogationVisitor =
+                    new ConstantPropogationVisitor(var_map);
+                    expression.accept(propogationVisitor);
+                    log.log(Level.FINEST, "\n__getExpression__\n " +
+                    propogationVisitor.getLogStr());
+                    expression = propogationVisitor.getExpression();
+                    log.log(Level.FINEST, "\n__getExpression__\n " +
+                    propogationVisitor.getLogStr());
+
+                    log.log(Level.FINEST, "After ConstantPropogation: " + expression);
+        			return expression;
+               } catch (VisitorException x) {
+                   log.log(Level.SEVERE,
+       					"encountered an exception -- this should not be happening!",
+       					x);
+               }
+               return null;
+           }
+
+/*
+ * Goal is to obtain a HashMap of variables and constant values assigned in the
+ * expression
+ */
+private static class MapVisitor extends Visitor {
+   private Stack<Expression> stack;
+   private String logstr;
+   private HashMap <Variable, Constant> map;
+
+   public MapVisitor() {
+	   stack = new Stack<Expression>();
+       map = new HashMap <Variable, Constant>();
+       logstr = "\n";
+   }
+
+   public HashMap getMap() {
+       logstr = "\n __Stack__\n";
+       while (!stack.isEmpty()) {
+           Expression x = stack.pop();
+
+           if (x instanceof Variable) {
+               //int c = ((IntConstant) map.get(var)).getValue();
```

Comment:
This method is quite busy, some inline comments would help the readability quite a bit.

Reply:
@PhillipVH , Please see new comments added for readability in latest commit. Will implement Javadoc commenting next time. Many thanks!